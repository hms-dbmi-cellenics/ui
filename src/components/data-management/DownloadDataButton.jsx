import React, { useEffect, useState } from 'react';
import _ from 'lodash';
import {
  Tooltip, Dropdown, Button, Space,
} from 'antd';

import { useSelector, useDispatch } from 'react-redux';
import { saveAs } from 'file-saver';
import { ClipLoader } from 'react-spinners';

import downloadTypes from 'utils/data-management/downloadTypes';
import endUserMessages from 'utils/endUserMessages';
import pipelineStatus from 'utils/pipelineStatusValues';
import { exportQCParameters, filterQCParameters } from 'utils/data-management/exportQCParameters';

import { loadBackendStatus } from 'redux/actions/backendStatus/index';

import { getBackendStatus } from 'redux/selectors';
import handleError from 'utils/http/handleError';
import downloadProcessedMatrix from 'utils/extraActionCreators/downloadProcessedMatrix';

const DownloadDataButton = () => {
  const dispatch = useDispatch();
  const experimentSettings = useSelector((state) => state.experimentSettings);
  const experiments = useSelector((state) => state.experiments);
  const { activeExperimentId } = experiments.meta;
  const activeExperiment = experiments[activeExperimentId];

  const {
    status: backendStatuses, loading: backendLoading,
  } = useSelector(getBackendStatus(activeExperimentId));

  const samples = useSelector((state) => state.samples);
  const [pipelineHasRun, setPipelineHasRun] = useState(false);
  const [allSamplesAnalysed, setAllSamplesAnalysed] = useState(false);
  const [downloadingProcessedSeurat, setDownloadingProcessedSeurat] = useState(false);
  const [dropdownExpanded, setDropdownExpanded] = useState(false);

  useEffect(() => {
    if (activeExperimentId && !backendLoading && !backendStatuses) {
      dispatch(loadBackendStatus(activeExperimentId));
    }
  }, [activeExperimentId]);

  useEffect(() => {
    setPipelineHasRun(
      activeExperimentId && (backendStatuses?.pipeline?.status === pipelineStatus.SUCCEEDED
        || backendStatuses?.seurat?.status === pipelineStatus.SUCCEEDED),
    );
  }, [backendStatuses]);

  useEffect(() => {
    setAllSamplesAnalysed(getAllSamplesAnalysed());
  }, [activeExperiment, experimentSettings]);

  const getAllSamplesAnalysed = () => {
    // Returns true only if there is at least one sample in the currently active
    // project AND all samples in the project have been analysed.
    if (!activeExperiment?.sampleIds?.length) {
      return false;
    }
    const steps = Object.values(_.omit(experimentSettings?.processing, ['meta']));
    return steps.length > 0
      // eslint-disable-next-line no-prototype-builtins
      && activeExperiment?.sampleIds?.every((s) => steps[0].hasOwnProperty(s));
  };

  const downloadExperimentData = async (type) => {
    try {
      if (!activeExperimentId) throw new Error('No experimentId specified');
      if (!downloadTypes.has(type)) throw new Error('Invalid download type');

      setDownloadingProcessedSeurat(true);
      await dispatch(downloadProcessedMatrix(activeExperimentId));
      setDownloadingProcessedSeurat(false);
      setDropdownExpanded(false);
    } catch (e) {
      handleError(e, endUserMessages.ERROR_DOWNLOADING_DATA);
    }
  };

  const menuItems = [
    {
      key: 'download-processed-seurat',
      disabled: !pipelineHasRun || backendLoading,
      onClick: (e) => {
        e.domEvent.stopPropagation();
        downloadExperimentData('processed-matrix');
      },
      label: (
        <Tooltip
          title={
            pipelineHasRun
              ? 'With Data Processing filters and settings applied'
              : 'Launch analysis to process data'
          }
          placement='left'
        >
          <Space>
            Processed Seurat object (.rds)
            {downloadingProcessedSeurat && <ClipLoader size={20} color='#8f0b10' />}
          </Space>
        </Tooltip>
      ),
    },
    {
      key: 'download-processing-settings',
      disabled: !allSamplesAnalysed || backendLoading,
      onClick: () => {
        const config = _.omit(experimentSettings.processing, ['meta']);
        const filteredConfig = filterQCParameters(
          config, activeExperiment.sampleIds, samples,
        );
        const blob = exportQCParameters(filteredConfig);
        saveAs(blob, `${activeExperimentId.split('-')[0]}_settings.txt`);
      },
      label: (
        allSamplesAnalysed
          ? 'Data Processing settings (.txt)'
          : (
            <Tooltip title='One or more of your samples has yet to be analysed' placement='left'>
              Data Processing settings (.txt)
            </Tooltip>
          )
      ),
    },
  ];

  return (
    <Dropdown
      open={dropdownExpanded}
      onOpenChange={(visible) => setDropdownExpanded(visible)}
      trigger={['click']}
      menu={{
        items: menuItems,
        onClick: (e) => {
          if (e.key !== 'download-processed-seurat') setDropdownExpanded(false);
        },
      }}
      placement='bottomRight'
      disabled={
        experiments.ids.length === 0
        || activeExperiment.sampleIds.length === 0
      }
    >
      <Button>
        Download
      </Button>
    </Dropdown>

  );
};

export default React.memo(DownloadDataButton);
