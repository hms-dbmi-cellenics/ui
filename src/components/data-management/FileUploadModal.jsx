import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import {
  Modal,
  Button,
  Typography,
  Select,
  Space,
  Row,
  Col,
  Empty,
  Divider,
  List,
} from 'antd';
import { CheckCircleTwoTone, CloseCircleTwoTone, DeleteOutlined } from '@ant-design/icons';
import Dropzone from 'react-dropzone';
import techOptions from 'utils/upload/fileUploadSpecifications';
import handleError from 'utils/http/handleError';
import { fileObjectToFileRecord } from 'utils/upload/processUpload';

import integrationTestConstants from 'utils/integrationTestConstants';
import endUserMessages from 'utils/endUserMessages';

const { Text, Title, Paragraph } = Typography;
const { Option } = Select;

const FileUploadModal = (props) => {
  const { onUpload, onCancel } = props;

  const guidanceFileLink = 'https://drive.google.com/file/d/1VPaB-yofuExinY2pXyGEEx-w39_OPubO/view';

  const [selectedTech, setSelectedTech] = useState('10X Chromium');
  const [canUpload, setCanUpload] = useState(false);
  const [filesList, setFilesList] = useState([]);

  useEffect(() => {
    setCanUpload(filesList.length && filesList.every((file) => file.valid));
  }, [filesList]);

  // Handle on Drop
  const onDrop = async (acceptedFiles) => {
    let filesNotInFolder = false;
    const filteredFiles = acceptedFiles
      // Remove all hidden files
      .filter((file) => !file.name.startsWith('.') && !file.name.startsWith('__MACOSX'))
      // Remove all files that aren't in a folder
      .filter((file) => {
        const inFolder = file.path.includes('/');

        filesNotInFolder ||= !inFolder;

        return inFolder;
      });

    if (filesNotInFolder) {
      handleError('error', endUserMessages.ERROR_FILES_FOLDER);
    }

    const newFiles = await Promise.all(filteredFiles.map((file) => (
      fileObjectToFileRecord(file, selectedTech)
    )));

    setFilesList([...filesList, ...newFiles]);
  };

  const removeFile = (fileName) => {
    const newArray = _.cloneDeep(filesList);

    const fileIdx = newArray.findIndex((file) => file.name === fileName);
    newArray.splice(fileIdx, 1);
    setFilesList(newArray);
  };

  const renderHelpText = () => (
    <>
      <Space direction='vertical' style={{ width: '100%' }}>
        <Paragraph>
          For each sample, upload a folder containing the
          {' '}
          {techOptions[selectedTech].inputInfo.length}
          {' '}
          count matrix files. The folder&apos;s name will be used to name the sample in it. You can change this name later in Data Management.
        </Paragraph>
        <Paragraph>
          The required files for each sample are:
        </Paragraph>
        <List
          dataSource={techOptions[selectedTech].inputInfo}
          size='small'
          itemLayout='vertical'
          bordered
          renderItem={(item) => (
            <List.Item>
              {
                item.map((fileName, i) => (
                  <span key={fileName}>
                    <Text code>{`${fileName}`}</Text>
                    {i !== item.length - 1 && ' or '}
                  </span>
                ))
              }
            </List.Item>
          )}
        />
      </Space>
    </>
  );

  return (
    <Modal
      title=''
      visible
      onCancel={onCancel}
      width='50%'
      footer={(
        <Button
          data-test-id={integrationTestConstants.ids.FILE_UPLOAD_BUTTON}
          type='primary'
          key='create'
          block
          disabled={!canUpload}
          onClick={() => {
            onUpload(filesList, selectedTech);
            setFilesList([]);
          }}
        >
          Upload
        </Button>
      )}
    >
      <Row>
        <Col span={24}>
          <Space direction='vertical' style={{ width: '100%' }}>
            <Space align='baseline'>
              <Title level={4} style={{ display: 'inline-block' }}>
                Technology:
                <span style={{ color: 'red', marginRight: '2em' }}>*</span>
              </Title>
              <Select
                defaultValue={selectedTech}
                onChange={(value) => setSelectedTech(value)}
              >
                {Object.keys(techOptions).map((val) => (
                  <Option key={`key-${val}`} value={val}>{val}</Option>
                ))}
              </Select>
            </Space>
            <Text type='secondary'>
              <i>
                Is your dataset generated using another single cell RNA-seq technology (e.g. Nadia, BD Rhapsody, etc.)? Email us to find out if we can support your data:
                <a href='mailto:hello@biomage.net'> hello@biomage.net</a>
              </i>
            </Text>
          </Space>
        </Col>
      </Row>

      <Row style={{ margin: '1rem 0' }}>
        <Col span={24}>
          <Title level={4}>
            File Upload:
            <span style={{ color: 'red', marginRight: '2em' }}>*</span>
          </Title>
          {selectedTech && renderHelpText()}
        </Col>
      </Row>

      <Row>
        {/* eslint-disable react/jsx-props-no-spreading */}
        <Col span={24}>
          <Paragraph type='secondary'>
            <i>
              Donâ€™t have the data in the accepted format? Email us for help with file conversion (e.g. from Fastq or H5 file):
              <a href='mailto:hello@biomage.net'> hello@biomage.net</a>
            </i>
            <span style={{ display: 'block', height: '0.6rem' }} />
            <i>
              More guidance on supported file types and formats is available
              <a rel='noreferrer' target='_blank' href={guidanceFileLink}> here</a>
              .
            </i>
          </Paragraph>
          <Dropzone onDrop={onDrop} multiple>
            {({ getRootProps, getInputProps }) => (
              <div
                data-test-id={integrationTestConstants.ids.FILE_UPLOAD_DROPZONE}
                style={{ border: '1px solid #ccc', padding: '2rem 0' }}
                {...getRootProps({ className: 'dropzone' })}
                id='dropzone'
              >
                <input data-test-id={integrationTestConstants.ids.FILE_UPLOAD_INPUT} {...getInputProps()} webkitdirectory='' />
                <Empty description='Drag and drop folders here or click to browse.' image={Empty.PRESENTED_IMAGE_SIMPLE} />
              </div>
            )}
          </Dropzone>
        </Col>
        {/* eslint-enable react/jsx-props-no-spreading */}

        {filesList.length ? (
          <>
            <Divider orientation='center'>To upload</Divider>
            <List
              dataSource={filesList}
              size='small'
              itemLayout='horizontal'
              grid='{column: 4}'
              renderItem={(file) => (

                <List.Item
                  key={file.name}
                  style={{ width: '100%' }}
                >
                  <Space>
                    {file.valid
                      ? (
                        <>
                          <CheckCircleTwoTone twoToneColor='#52c41a' />
                        </>
                      ) : (
                        <>
                          <CloseCircleTwoTone twoToneColor='#f5222d' />
                        </>
                      )}
                    <Text
                      ellipsis={{ tooltip: file.name }}
                      style={{ width: '200px' }}
                    >
                      {file.name}

                    </Text>
                    <DeleteOutlined style={{ color: 'crimson' }} onClick={() => { removeFile(file.name); }} />
                  </Space>
                </List.Item>

              )}
            />
          </>
        ) : ''}
      </Row>
    </Modal>

  );
};

FileUploadModal.propTypes = {
  onUpload: PropTypes.func,
  onCancel: PropTypes.func,
};

FileUploadModal.defaultProps = {
  onUpload: null,
  onCancel: null,
};

export default FileUploadModal;
