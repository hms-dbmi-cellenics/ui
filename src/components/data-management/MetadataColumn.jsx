import React from 'react';
import { Space, Input } from 'antd';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { DEFAULT_NA } from 'redux/reducers/projects/initialState';
import {
  updateMetadataTrack,
} from 'redux/actions/projects';
import validateInputs, { rules } from 'utils/validateInputs';
import { metadataNameToKey } from 'utils/data-management/metadataUtils';
import EditableField from '../EditableField';
import MetadataEditor from './MetadataEditor';

const MetadataColumnTitle = (props) => {
  const {
    name, sampleNames, activeProjectUuid, deleteMetadataColumn, setCells,
  } = props;

  const validationParams = {
    existingNames: sampleNames,
  };

  return (
    <MetadataTitle
      name={name}
      validateInput={
        (newName, metadataNameValidation) => validateInputs(
          newName, metadataNameValidation, validationParams,
        ).isValid
      }
      setCells={setCells}
      deleteMetadataColumn={deleteMetadataColumn}
      activeProjectUuid={activeProjectUuid}
    />
  );
};

MetadataColumnTitle.propTypes = {
  name: PropTypes.string.isRequired,
  sampleNames: PropTypes.instanceOf(Set).isRequired,
  setCells: PropTypes.func.isRequired,
  deleteMetadataColumn: PropTypes.func.isRequired,
  activeProjectUuid: PropTypes.string.isRequired,
};

const MetadataTitle = (props) => {
  const dispatch = useDispatch();
  const {
    name, validateInput, setCells, deleteMetadataColumn, activeProjectUuid,
  } = props;
  const metaKey = metadataNameToKey(name);

  const metadataNameValidation = [
    rules.MIN_1_CHAR,
    rules.ALPHANUM_SPACE,
    rules.START_WITH_ALPHABET,
    rules.UNIQUE_NAME_CASE_INSENSITIVE,
  ];
  return (
    <Space>
      <EditableField
        deleteEnabled
        onDelete={(e, currentName) => deleteMetadataColumn(currentName)}
        onAfterSubmit={(newName) => dispatch(
          updateMetadataTrack(name, newName, activeProjectUuid),
        )}
        value={name}
        validationFunc={
          (newName) => validateInput(newName, metadataNameValidation)
        }
      />
      <MetadataEditor
        onReplaceEmpty={(value) => setCells(value, metaKey, 'REPLACE_EMPTY')}
        onReplaceAll={(value) => setCells(value, metaKey, 'REPLACE_ALL')}
        onClearAll={() => setCells(DEFAULT_NA, metaKey, 'CLEAR_ALL')}
        massEdit
      >
        <Input />
      </MetadataEditor>
    </Space>
  );
};
MetadataTitle.propTypes = {
  name: PropTypes.string.isRequired,
  validateInput: PropTypes.func.isRequired,
  setCells: PropTypes.func.isRequired,
  deleteMetadataColumn: PropTypes.func.isRequired,
  activeProjectUuid: PropTypes.string.isRequired,
};
export default MetadataColumnTitle;
