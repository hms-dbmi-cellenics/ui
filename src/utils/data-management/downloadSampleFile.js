import downloadFromUrl from 'utils/downloadFromUrl';
import fetchAPI from 'utils/http/fetchAPI';

const downloadSampleFile = async (experimentId, sampleUuid, fileType) => {
  const requestUrl = `/v2/experiments/${experimentId}/samples/${sampleUuid}/files/${fileType}/downloadUrl`;
  const downloadUrl = await fetchAPI(requestUrl);

  downloadFromUrl(downloadUrl);
};

export default downloadSampleFile;
