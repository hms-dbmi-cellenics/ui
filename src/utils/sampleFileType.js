const sampleFileType = {
  BARCODES_10_X: 'barcodes10x',
  FEATURES_10_X: 'features10x',
  MATRIX_10_X: 'matrix10x',
  H5_10_X: '10x_h5',
  SEURAT: 'seurat',
  RHAPSODY: 'rhapsody',
};

const fileTypeToDisplay = {
  [sampleFileType.BARCODES_10_X]: 'barcodes.tsv',
  [sampleFileType.FEATURES_10_X]: 'genes.tsv',
  [sampleFileType.MATRIX_10_X]: 'matrix.mtx',
  [sampleFileType.H5_10_X]: 'matrix.h5',
  [sampleFileType.SEURAT]: 'seurat rds',
  [sampleFileType.RHAPSODY]: 'expression_data.st',
};

export default sampleFileType;

export { fileTypeToDisplay };
