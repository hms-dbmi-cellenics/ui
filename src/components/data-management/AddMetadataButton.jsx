import React, { useState } from 'react';
import {
  Menu, Dropdown, Button,
} from 'antd';
import PropTypes from 'prop-types';

import { useSelector, useDispatch } from 'react-redux';

import uploadMetadataFile from 'redux/actions/experiments/uploadMetadataFile';
import { sampleTech } from 'utils/constants';
import MetadataUploadModal from './MetadataUploadModal';

const AddMetadataButton = ({ samplesTableRef }) => {
  const dispatch = useDispatch();
  const { activeExperimentId } = useSelector((state) => state.experiments.meta);
  const activeExperiment = useSelector((state) => state.experiments[activeExperimentId]);
  const isSubsetted = activeExperiment?.isSubsetted;
  const samples = useSelector((state) => state.samples);
  const selectedTech = samples[activeExperiment?.sampleIds[0]]?.type;
  const [uploadModalVisible, setUploadModalVisible] = useState(false);

  const uploadFiles = (file) => {
    dispatch(uploadMetadataFile(activeExperimentId, file));
    setUploadModalVisible(false);
  };

  return (
    <>
      <Dropdown
        overlay={() => (
          <Menu>
            <Menu.Item
              key='add-metadata-column'
              onClick={() => samplesTableRef.current.createMetadataColumn()}
            >
              Create track
            </Menu.Item>
            <Menu.Item
              key='upload-metadata-file'
              onClick={() => {
                setUploadModalVisible(true);
              }}
            >
              Upload file
            </Menu.Item>
          </Menu>
        )}
        trigger={['click']}
        placement='bottomRight'
        disabled={activeExperiment.sampleIds?.length === 0 || isSubsetted || selectedTech === sampleTech.SEURAT}
      >
        <Button>
          Add metadata
        </Button>
      </Dropdown>
      {uploadModalVisible && (
        <MetadataUploadModal
          onUpload={uploadFiles}
          onCancel={() => setUploadModalVisible(false)}
        />
      )}
    </>
  );
};

AddMetadataButton.propTypes = {
  samplesTableRef: PropTypes.shape({ current: PropTypes.any }).isRequired,
};

export default AddMetadataButton;

// Stub commit
