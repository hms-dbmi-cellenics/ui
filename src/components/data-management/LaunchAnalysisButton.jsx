import React, { useState, useCallback, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  Button, Tooltip, Popconfirm, Typography, Space,
} from 'antd';
import _ from 'lodash';

import { modules, sampleTech } from 'utils/constants';

import UploadStatus from 'utils/upload/UploadStatus';
import integrationTestConstants from 'utils/integrationTestConstants';

import { useAppRouter } from 'utils/AppRouteProvider';
import calculatePipelinesRerunStatus from 'utils/data-management/calculatePipelinesRerunStatus';
import { WarningOutlined } from '@ant-design/icons';

const { Text } = Typography;

const LaunchAnalysisButton = () => {
  const dispatch = useDispatch();
  const { navigateTo } = useAppRouter();

  const experiments = useSelector((state) => state.experiments);
  const samples = useSelector((state) => state.samples);
  const backendStatus = useSelector((state) => state.backendStatus);

  const { activeExperimentId } = experiments.meta;
  const activeExperiment = experiments[activeExperimentId];
  const selectedTech = samples[activeExperiment?.sampleIds[0]]?.type;
  const isTechSeurat = selectedTech === sampleTech.SEURAT;

  const [pipelinesRerunStatus, setPipelinesRerunStatus] = useState({
    runPipeline: null, rerun: true, reasons: [], complete: false,
  });

  const launchAnalysis = async () => {
    let shouldNavigate = true;
    if (pipelinesRerunStatus.rerun) {
      shouldNavigate = await dispatch(pipelinesRerunStatus.runPipeline(activeExperimentId));
    }

    if (shouldNavigate) {
      const moduleName = isTechSeurat && pipelinesRerunStatus.complete
        ? modules.DATA_EXPLORATION : modules.DATA_PROCESSING;
      navigateTo(moduleName, { experimentId: activeExperimentId });
    }
  };

  useEffect(() => {
    // The value of backend status is null for new experiments that have never run
    const setupPipeline = isTechSeurat ? 'seurat' : 'gem2s';
    const {
      pipeline: qcBackendStatus, [setupPipeline]: setupBackendStatus,
    } = backendStatus[activeExperimentId]?.status ?? {};

    if (
      !setupBackendStatus
      || !experiments[activeExperimentId]?.sampleIds?.length > 0
    ) return;

    setPipelinesRerunStatus(
      calculatePipelinesRerunStatus(
        setupBackendStatus,
        qcBackendStatus,
        activeExperiment,
        isTechSeurat,
      ),
    );
  }, [backendStatus, activeExperimentId, samples, activeExperiment]);

  const cellLevelMetadataIsReady = (
    _.isNil(activeExperiment.cellLevelMetadata)
    || activeExperiment.cellLevelMetadata.uploadStatus === UploadStatus.UPLOADED
  );

  const canLaunchAnalysis = useCallback(() => {
    if (activeExperiment.sampleIds.length === 0) return false;

    // Check that samples is loaded
    const testSampleUuid = activeExperiment.sampleIds[0];
    if (samples[testSampleUuid] === undefined) return false;

    const metadataKeysAvailable = activeExperiment.metadataKeys.length;

    const allSampleFilesUploaded = (sample) => (
      Object.values(sample.files).every((file) => file.upload.status === UploadStatus.UPLOADED)
    );

    const allSampleMetadataInserted = (sample) => {
      if (!metadataKeysAvailable) return true;

      if (Object.keys(sample.metadata).length !== metadataKeysAvailable) return false;
      return Object.values(sample.metadata)
        .every((value) => value.length > 0);
    };

    const canLaunch = activeExperiment.sampleIds.every((sampleUuid) => {
      if (!samples[sampleUuid]) return false;

      const checkedSample = samples[sampleUuid];

      return allSampleFilesUploaded(checkedSample)
        && allSampleMetadataInserted(checkedSample);
    }) && cellLevelMetadataIsReady;

    return canLaunch;
  }, [
    samples,
    activeExperiment?.sampleIds,
    activeExperiment?.metadataKeys,
    activeExperiment.cellLevelMetadata?.uploadStatus,
  ]);

  const getAllExperimentSampleFiles = useCallback(() => (
    activeExperiment.sampleIds.flatMap((sampleId) => Object.values(samples[sampleId]?.files ?? {}))
  ), [samples, activeExperiment?.sampleIds]);

  const getAnyFileUploadFailed = useCallback(() => {
    const nonErrorStatuses = [UploadStatus.UPLOADED, UploadStatus.UPLOADING];

    const cellLevelUploadError = !_.isNil(activeExperiment.cellLevelMetadata)
      && !nonErrorStatuses.includes(activeExperiment.cellLevelMetadata.uploadStatus);

    const sampleFileUploadError = getAllExperimentSampleFiles()
      .some((sampleFile) => !nonErrorStatuses.includes(sampleFile.upload.status));

    return cellLevelUploadError || sampleFileUploadError;
  }, [
    samples,
    activeExperiment?.sampleIds,
    activeExperiment.cellLevelMetadata,
  ]);

  const LaunchButtonTemplate = (props) => {
    const {
      // eslint-disable-next-line react/prop-types
      onClick, disabled, text, loading,
    } = props;

    return (
      <Button
        data-test-id={integrationTestConstants.ids.PROCESS_PROJECT_BUTTON}
        type='primary'
        disabled={disabled}
        onClick={onClick}
        loading={loading}
      >
        {
          getAnyFileUploadFailed() ? (
            <Space>
              <Text type='danger'>
                <WarningOutlined />
              </Text>
              {text}
            </Space>
          ) : (
            text
          )
        }

      </Button>
    );
  };

  const renderLaunchButton = () => {
    let buttonText;

    if (pipelinesRerunStatus.rerun) {
      buttonText = 'Process project';
    } else if (isTechSeurat && pipelinesRerunStatus.complete) {
      buttonText = 'Go to Data Exploration';
    } else {
      buttonText = 'Go to Data Processing';
    }

    if (!backendStatus[activeExperimentId] || backendStatus[activeExperimentId]?.loading) {
      return <LaunchButtonTemplate text='Loading project...' disabled loading />;
    }

    if (!canLaunchAnalysis()) {
      const message = !cellLevelMetadataIsReady
        ? 'Ensure that the cell level metadata file is uploaded correctly'
        : 'Ensure that all samples are uploaded successfully and all relevant metadata is inserted.';
      return (
        <Tooltip
          title={message}
        >
          {/* disabled button inside tooltip causes tooltip to not function */}
          {/* https://github.com/react-component/tooltip/issues/18#issuecomment-140078802 */}
          <span>
            <LaunchButtonTemplate text={buttonText} disabled />
          </span>
        </Tooltip>
      );
    }

    if (pipelinesRerunStatus.rerun) {
      return (
        <Popconfirm
          title={`This project has to be processed because ${pipelinesRerunStatus.reasons.join(' and ')}. \
        This will take several minutes.\
        Do you want to continue?`}
          onConfirm={() => launchAnalysis()}
          okText='Yes'
          okButtonProps={{ 'data-test-id': integrationTestConstants.ids.CONFIRM_PROCESS_PROJECT }}
          cancelText='No'
          placement='bottom'
          overlayStyle={{ maxWidth: '250px' }}
        >
          <LaunchButtonTemplate text={buttonText} />
        </Popconfirm>
      );
    }

    return <LaunchButtonTemplate text={buttonText} onClick={() => launchAnalysis()} />;
  };

  return renderLaunchButton();
};

export default LaunchAnalysisButton;
