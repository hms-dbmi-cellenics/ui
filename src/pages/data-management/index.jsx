import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { loadExperiments } from 'redux/actions/experiments';

import Header from 'components/Header';
import MultiTileContainer from 'components/MultiTileContainer';
import NewProjectModal from 'components/data-management/NewProjectModal';
import ProjectsListContainer from 'components/data-management/project/ProjectsListContainer';
import ProjectDetails from 'components/data-management/project/ProjectDetails';
import { loadProcessingSettings } from 'redux/actions/experimentSettings';
import loadBackendStatus from 'redux/actions/backendStatus/loadBackendStatus';
import { loadSamples } from 'redux/actions/samples';
import ExampleExperimentsSpace from 'components/data-management/ExampleExperimentsSpace';
import Loader from 'components/Loader';
import { privacyPolicyIsNotAccepted } from 'utils/deploymentInfo';

const DataManagementPage = () => {
  const dispatch = useDispatch();

  const samples = useSelector((state) => state.samples);

  const { activeExperimentId } = useSelector((state) => state.experiments.meta);
  const experiments = useSelector(((state) => state.experiments));
  const user = useSelector((state) => state.user.current);

  const activeExperiment = experiments[activeExperimentId];
  const domainName = useSelector((state) => state.networkResources?.domainName);

  const [newProjectModalVisible, setNewProjectModalVisible] = useState(false);

  useEffect(() => {
    if (privacyPolicyIsNotAccepted(user, domainName)) return;
    if (experiments.ids.length === 0) dispatch(loadExperiments());
  }, [user]);

  const samplesAreLoaded = () => {
    const loadedSampleIds = Object.keys(samples);
    return activeExperiment.sampleIds.every((sampleId) => loadedSampleIds.includes(sampleId));
  };

  useEffect(() => {
    // If the active experiment isnt loaded, reload
    if (activeExperimentId && !activeExperiment) {
      dispatch(loadExperiments());
    }
  }, [activeExperiment]);

  useEffect(() => {
    if (!activeExperimentId
      || !activeExperiment
       || privacyPolicyIsNotAccepted(user, domainName)
    ) return;

    dispatch(loadProcessingSettings(activeExperimentId));

    if (!samplesAreLoaded()) dispatch(loadSamples(activeExperimentId));

    dispatch(loadBackendStatus(activeExperimentId));
  }, [activeExperimentId, activeExperiment, user]);

  const PROJECTS_LIST = 'Projects';
  const PROJECT_DETAILS = 'Project Details';

  const TILE_MAP = {
    [PROJECTS_LIST]: {
      toolbarControls: [],
      component: (width, height) => (
        <ProjectsListContainer
          height={height}
          onCreateNewProject={() => setNewProjectModalVisible(true)}
        />
      ),
    },
    [PROJECT_DETAILS]: {
      toolbarControls: [],
      component: (width, height) => {
        if (!activeExperimentId) {
          return <ExampleExperimentsSpace introductionText='You have no projects yet.' />;
        }

        if (!activeExperiment) {
          return (
            <center>
              <Loader />
            </center>
          );
        }

        return (
          <ProjectDetails
            width={width}
            height={height}
          />
        );
      },
    },
  };

  const windows = {
    direction: 'row',
    first: PROJECTS_LIST,
    second: PROJECT_DETAILS,
    splitPercentage: 23,
  };

  return (
    <>
      <Header title='Data Management' />
      {newProjectModalVisible ? (
        <NewProjectModal
          onCancel={() => { setNewProjectModalVisible(false); }}
          onCreate={() => { setNewProjectModalVisible(false); }}
        />
      ) : (<></>)}
      <MultiTileContainer
        tileMap={TILE_MAP}
        initialArrangement={windows}
      />
    </>
  );
};

export default DataManagementPage;
