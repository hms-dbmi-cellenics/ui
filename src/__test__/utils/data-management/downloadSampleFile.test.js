import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import { downloadSampleFile } from 'utils/data-management/downloadSampleFile';
import downloadFromUrl from 'utils/downloadFromUrl';

import fake from '__test__/test-utils/constants';
import sampleFileType from 'utils/sampleFileType';

jest.mock('utils/downloadFromUrl');

enableFetchMocks();

describe('downloadFromUrl', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    fetchMock.resetMocks();
    fetchMock.doMock();
  });

  it('Downloads from url', async () => {
    const mockSignedUrl = [{ url: 'mockDownloadUrl', fileId: 'fake-file-id' }];

    fetchMock.mockResponse(JSON.stringify(mockSignedUrl));

    await downloadSampleFile(fake.EXPERIMENT_ID, fake.SAMPLE_ID, sampleFileType.FEATURES_10_X);

    expect(downloadFromUrl).toHaveBeenCalledWith(mockSignedUrl[0].url);
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });
});
