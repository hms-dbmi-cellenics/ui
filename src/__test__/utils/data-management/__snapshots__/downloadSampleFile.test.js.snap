// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`downloadFromUrl Downloads from url 1`] = `
[
  [
    "http://localhost:3000/v2/experiments/testae48e318dab9a1bd0bexperiment/samples/test9188-d682-test-mock-cb6d644cmock/files/features10x/downloadUrl",
    {
      "headers": {},
    },
  ],
]
`;
